lib/clients/public.js

```diff
@@ -1,206 +1,214 @@
-var request = require('request');
-var _ = {
-  'forEach': require('lodash.foreach'),
-  'assign': require('lodash.assign'),
-  'partial': require('lodash.partial'),
-};
-
-var Readable = require('stream').Readable;
-
-var PublicClient = function(productID, apiURI) {
-  var self = this;
-  self.productID = productID || 'BTC-USD';
-  self.apiURI = apiURI || 'https://api.gdax.com';
-};
-
-var API_LIMIT = 100;
+const request = require('request');
+const { Readable } = require('stream');
+
+class PublicClient {
+  constructor(productID = 'BTC-USD', apiURI = 'https://api.gdax.com') {
+    this.productID = productID;
+    this.apiURI = apiURI;
+    this.API_LIMIT = 100;
+  }
 
-_.assign(PublicClient.prototype, new function() {
-  var prototype = this;
+  get(...args) {
+    return this.request('get', ...args);
+  }
+  put(...args) {
+    return this.request('put', ...args);
+  }
+  post(...args) {
+    return this.request('post', ...args);
+  }
+  delete(...args) {
+    return this.request('delete', ...args);
+  }
 
-  prototype.addHeaders = function(obj, additional) {
+  addHeaders(obj, additional) {
     obj.headers = obj.headers || {};
-    return _.assign(obj.headers, {
-      'User-Agent': 'gdax-node-client',
-      'Accept': 'application/json',
-      'Content-Type': 'application/json',
-    }, additional);
-  };
-
-  prototype.makeRelativeURI = function(parts) {
+    return Object.assign(
+      obj.headers,
+      {
+        'User-Agent': 'gdax-node-client',
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      additional
+    );
+  }
+
+  makeRelativeURI(parts) {
     return '/' + parts.join('/');
-  };
+  }
 
-  prototype.makeAbsoluteURI = function(relativeURI) {
+  makeAbsoluteURI(relativeURI) {
     return this.apiURI + relativeURI;
-  };
+  }
 
-  prototype.makeRequestCallback = function(callback) {
+  makeRequestCallback(callback, resolve, reject) {
     return function(err, response, data) {
       try {
         data = JSON.parse(data);
       } catch (e) {
-        data = null
+        data = null;
       }
-      callback(err, response, data);
+      if (typeof callback == 'function') {
+        callback(err, response, data);
+        return;
+      }
+      if (err) reject(err);
+      if (data === null)
+        reject(new Error('Response could not be parsed as JSON'));
+      else resolve(data);
```

Comment:
I've been looking into replacing parts of this library with [swagger-codegen](https://github.com/swagger-api/swagger-codegen). Their [promise interface](https://github.com/swagger-api/swagger-codegen/blob/0416b0332f36b4a7f80a4313df8f8257a4539fe4/modules/swagger-codegen/src/main/resources/Javascript/es6/ApiClient.mustache#L462) returns a `{data, response}` object — I guess we should adopt that as well. While I like the simplicity of this API, there probably are cases where you want the request object (for headers, status codes etc.), plus an object would allow for future extensions.